include/swift/AST/IRGenRequests.h

```diff
@@ -0,0 +1,92 @@
+//===--- IRGenRequests.h - IRGen Requests -----------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines IRGen requests.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_IRGen_REQUESTS_H
+#define SWIFT_IRGen_REQUESTS_H
+
+#include "swift/AST/ASTTypeIDs.h"
+#include "swift/AST/SimpleRequest.h"
+#include "swift/Basic/PrimarySpecificPaths.h"
+#include "llvm/ADT/StringSet.h"
+
+namespace swift {
+class SourceFile;
+namespace irgen {
+  class IRGenModule;
+};
+
+}; // namespace swift
+
+namespace llvm {
+class GlobalVariable;
+class LLVMContext;
+class Module;
+}; // namespace llvm
+
+namespace swift {
+
+
+/// Report that a request of the given kind is being evaluated, so it
+/// can be recorded by the stats reporter.
+template<typename Request>
+void reportEvaluatedRequest(UnifiedStatsReporter &stats,
+                            const Request &request);
+
+class IRGenSourceFileRequest
+    : public SimpleRequest<IRGenSourceFileRequest,
+                           bool(irgen::IRGenModule *, SourceFile *),
+                           CacheKind::Uncached> {
+public:
+  using SimpleRequest::SimpleRequest;
+
+private:
+  friend SimpleRequest;
+
+  // Evaluation.
+  llvm::Expected<bool>
+  evaluate(Evaluator &evaluator, irgen::IRGenModule *IGM, SourceFile *SF) const;
```

Comment:
As with the SILGen PR, my feedback here is that the IGM should be internal to the request itself. Again, you might consider splitting up the WMO and non-WMO code paths into separate requests because the scope of the IGM instance is different in each case.